### PR TITLE
Move to the new way to do Redux

### PR DESCRIPTION
**Pull request by Nepomuceno**
### What does this PR do?

Hey there! 🌟 This PR is a big one, and I'm super excited about it! 🥳 The main goal here is to bring our application up to speed by migrating to a modern version of Redux. It's an important step because, down the road, we plan to transition some functionality to React Query. But before we can do that, we need to make sure we're following the latest React practices.

As part of this update, we made a significant change. We've stopped depending on the window component for `iqdata` and `fftdata`, and now we store all of this data in Redux. 🔄 This ensures a more centralized and organized approach to managing our application state.

In the process, we also identified some performance issues that needed attention. 🚀

To give everyone a chance to review the code and test it locally, I've opened the PR as a draft. Your feedback and input would be invaluable!

By the way, if you spot anything strange or out of place in the code, please feel free to make corrections. 🙏 Your help would be much appreciated! Let's make this update a success! 🎉

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you successfully ran tests with your changes locally? Including container validation.
* [x] Have you lint your code locally prior to submission?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you labelled this PR correctly? E.G. major, milestone, breaking, breaking-change, minor, feature, enhancement, patch, fix, chore, refactor etc

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

### CODE REVIEW

**Comment by Nepomuceno** [comment]
Now I got Spectogram kind of working :) would love some extra help to smash the latest bugs that I found on it :) 
I also removed all the window.fftData and window.iqData and moved everything to redux 😄

**Comment by Nepomuceno** [comment]
Rulers are now fixed 👍

File: client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx [hunk 1]
```diff
@@ -102,20 +104,24 @@ export default function RecordingsBrowser(props) {
...
       const accountName = searchParams.get('accountName');
       const containerName = searchParams.get('containerName');
       const sasToken = searchParams.get('sasToken');
-      updateConnectionAccountName(accountName);
-      updateConnectionContainerName(containerName);
-      updateConnectionSasToken(sasToken);
-      fetchRecordingsList({ accountName: accountName, containerName: containerName, sasToken: sasToken });
+      dispatch(updateConnectionAccountName(accountName));
+      dispatch(updateConnectionContainerName(containerName));
+      dispatch(updateConnectionSasToken(sasToken));
```
Comment (Jaydee123) [inline client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx:115]: This makes me spicy

File: client/src/Components/Spectrogram/AnnotationViewer.jsx [hunk 2]
```diff
@@ -25,17 +30,20 @@ const AnnotationViewer = (props) => {
...
       (annotations[annot_indx].y2 - annotations[annot_indx].y1) * fftSize * zoomLevel;
     let lower_freq = meta.captures[0]['core:frequency'] - meta.global['core:sample_rate'] / 2;
-    updatedAnnotations[f]['core:freq_lower_edge'] =
+    updatedAnnotation['core:freq_lower_edge'] =
       (annotations[annot_indx].x1 / fftSize) * meta.global['core:sample_rate'] + lower_freq;
-    updatedAnnotations[f]['core:freq_upper_edge'] =
+    updatedAnnotation['core:freq_upper_edge'] =
       (annotations[annot_indx].x2 / fftSize) * meta.global['core:sample_rate'] + lower_freq;
-    props.handleMeta(updatedAnnotations);
+    console.log(updatedAnnotation);
```
Comment (Jaydee123) [inline client/src/Components/Spectrogram/AnnotationViewer.jsx:45]: Pls remove

File: client/src/Components/RepoBrowser/LocalFileBrowser.jsx [hunk 3]
```diff
@@ -40,29 +47,35 @@ async function handleDirectoryEntry(handle, out, dir) {
...
 
   const openFile = async () => {
+    dispatch(clearRecordingsList());
+    dispatch(resetConnection());
     const [handle1, handle2] = await window.showOpenFilePicker({ multiple: true });
     const file1 = await handle1.getFile();
     if (file1.name.includes('.sigmf-meta')) {
-      props.updateConnectionMetaFileHandle(handle1); // store it in redux
-      props.updateConnectionDataFileHandle(handle2); // assume other file is data
+      dispatch(updateConnectionMetaFileHandle(handle1)); // store it in redux
```
[the remark this answers is not in the document]
Comment (Nepomuceno) [reply client/src/Components/RepoBrowser/LocalFileBrowser.jsx:62]: would make sense naming things it is hard but we dont knwo in advance whicxh one will be the metadata handler and which one will be the the data handler.

File: client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx (hunk 1 above)
Comment (Nepomuceno) [reply client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx:115]: It is the current way that is working we hope to change this soon

File: client/src/Components/Spectrogram/AnnotationViewer.jsx (hunk 2 above)
Comment (Nepomuceno) [reply client/src/Components/Spectrogram/AnnotationViewer.jsx]: Fixed

**Comment by Nepomuceno** [comment]
I think all the bugs have been correct this should be ready for prime time 🥳

**Comment by 777arc** [comment]
It looks like the spectrogram doesnt load until after the minimap fetching is done, which could mean quite a bit of delay for folks with a slow connection, it would be nice if the spectrogram would display its initial content immediately like it used to be

**Comment by Nepomuceno** [comment]
> 



> It looks like the spectrogram doesnt load until after the minimap fetching is done, which could mean quite a bit of delay for folks with a slow connection, it would be nice if the spectrogram would display its initial content immediately like it used to be

Did fix that now the image it is fetched on the start of the page.

File: client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx [hunk 4]
```diff
@@ -35,6 +45,7 @@ function GroupByFolder(files, root) {
       if (isAFile) {
         currentFolder.contents.push({
           ...file,
+          relativeKey: relativeKey,
```
Comment (hannahlspencer) [inline client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx:48]: why not keep the file.relativeKey being instantiated as before?

File: client/src/Components/Spectrogram/ScrollBar.jsx [hunk 5]
```diff
@@ -53,95 +53,95 @@ const ScrollBar = (props) => {
...
-        f.transform(out, samples); // assumes input (2nd arg) is in form IQIQIQIQ and twice the length of fft.size
-        let magnitudes = new Array(out.length / 2);
-        for (let j = 0; j < out.length / 2; j++) {
-          magnitudes[j] = Math.sqrt(Math.pow(out[j * 2], 2) + Math.pow(out[j * 2 + 1], 2)); // take magnitude
-        }
-        fftshift(magnitudes); // in-place
-        magnitudes = magnitudes.map((x) => 10.0 * Math.log10(x)); // convert to dB
-        magnitudes = magnitudes.map((x) => (isFinite(x) ? x : 0)); // get rid of -infinity which happens when the input is all 0s
-        magnitudesBuffer.set(magnitudes, i * fftSizeScrollbar);
+    console.log(`minimap fetches complete ${minimap}`);
```
Comment (hannahlspencer) [inline client/src/Components/Spectrogram/ScrollBar.jsx:66]: necessary console.log?

File: client/src/Store/Reducers/FetchMetaReducer.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import { updateConnectionBlobClient } from './ConnectionReducer';
+
+const initialState = { annotations: [], captures: [], global: {} };
+
+export const fetchMeta = createAsyncThunk('fetchMata/fetchMeta', async (args: any, thunkAPI) => {
+  console.log('running fetchMeta');
+  const connection = args;
+  let meta_string = '';
+  let meta_json;
+  let blobName = connection.recording + '.sigmf-meta'; // has to go outside of condition or else react gets mad
```
Comment (hannahlspencer) [inline client/src/Store/Reducers/FetchMetaReducer.ts:18]: lol great comment

**Review by hannahlspencer — APPROVED** [review]

File: client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx (hunk 4 above)
Comment (Nepomuceno) [reply client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx:48]: Dont know will check :) dont tremember why I did this change anymore 😆

Comment (Nepomuceno) [reply client/src/Components/RecordingsBrowser/RecordingsBrowser.jsx:48]: Ok remember now it was not working for local folders becaue you needed to have the relative key as part of the current folder information.

File: client/src/Store/Reducers/BlobReducer.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+      state.pythonSnippet = action.payload;
+    },
+    updateBlobSize: (state, action: PayloadAction<number>) => {
+      state.size = action.payload;
+    },
+    updateBlobTotalIQSamples: (state, action: PayloadAction<number>) => {
+      state.totalIQSamples = action.payload;
+    },
+    // doesnt appear to be used?
+    //updateBlobIQData: (state, action: PayloadAction<any>) => {
```
[the remark this answers is not in the document]
Comment (Nepomuceno) [reply client/src/Store/Reducers/BlobReducer.ts:72]: Good catch will remove it !

File: client/src/Components/RepoBrowser/RepoBrowser.jsx [hunk 8]
```diff
@@ -11,7 +11,7 @@ import SiggenTile from './SiggenTile';
 import ValidatorTile from './ValidatorTile';
 import { configQuery } from '../../api/config/queries';
```
Comment (Jaydee123) [inline client/src/Components/RepoBrowser/RepoBrowser.jsx:13]: This class has the best change

File: client/src/Components/Spectrogram/AnnotationViewer.jsx [hunk 9]
```diff
@@ -25,17 +29,18 @@ const AnnotationViewer = (props) => {
 
     // Now update the actual meta.annotations
     const f = annotations[annot_indx]['index']; // remember there are 2 different indexes- the ones on the screen and the meta.annotations
-    let updatedAnnotations = [...props.meta.annotations];
+    let updatedAnnotations = [...meta.annotations];
+    const updatedAnnotation = { ...updatedAnnotations[f] };
```
Comment (Jaydee123) [inline client/src/Components/Spectrogram/AnnotationViewer.jsx:33]: Should this be a constant when you are making changes to it further diwn?

Comment (Nepomuceno) [reply client/src/Components/Spectrogram/AnnotationViewer.jsx:33]: Good catch will fix.

File: client/src/Components/RepoBrowser/RepoBrowser.jsx (hunk 8 above)
Comment (Nepomuceno) [reply client/src/Components/RepoBrowser/RepoBrowser.jsx:13]: a LOT smaller now :)

**Review by 777arc — APPROVED** [review]

**Pull request merged**